File: Leetcoding-Actions/LC-Monthly-Challenges/2020-11/15-Range-Sum-of-BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        '''
        T: O(n) and S: O(n)
        '''
        if not root: return 0
        
        sum = 0
        stack = []
        curr = root
        
        while True:
            while curr:
                stack.append(curr)
                curr = curr.left
            if not stack: break
                
            curr = stack.pop()
            
            if curr.val >= low and curr.val <= high:
                sum += curr.val
            if curr.val > high: return sum
            
            curr = curr.right
            
        return sum

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-11/15-Range-Sum-of-BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        /**
        * T: O(n) and S: O(n)
        */
        if (root == null) return 0;
        
        int sum = 0;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;
        
        while (true) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            if (stack.isEmpty()) break;
            
            curr = stack.pop();
            
            if (curr.val >= low && curr.val <= high) {
                sum += curr.val;
            }
            if (curr.val > high) {
                return sum;
            }
            
            curr = curr.right;
        }
        return sum;
    }
}
